File: android2_lesson03/src/java/java/com/example/ama/android2_lesson03/repo/data/location/LocationManagerAndroid.java
package com.example.ama.android2_lesson03.repo.data.location;

import android.content.Context;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.location.LocationProvider;
import android.os.Bundle;
import android.util.Log;

import com.example.ama.android2_lesson03.PocketMap;
import com.example.ama.android2_lesson03.R;
import com.example.ama.android2_lesson03.repo.data.base.BaseLocationManager;
import com.example.ama.android2_lesson03.utils.PermissionManager;

/**
 * Class for locating by android.location
 */
public class LocationManagerAndroid extends BaseLocationManager {

    private static final long REQUEST_MIN_TIME = 3000L;
    private static final float REQUEST_MIN_DISTANCE = 0;

    private LocationManager locManager;
    private LocationListener locListener;

    public LocationManagerAndroid() {
        super();
        locManager = (LocationManager) PocketMap.getInstance().getSystemService(Context.LOCATION_SERVICE);
    }

    /**
     * Find user location by Passive Provider and notify about result by callback
     *
     * @param callback for sending result
     */

    @Override
    public void findMyLocation(OnLocationSearchResultCallback callback) {
        if (PermissionManager.checkPermission(PocketMap.getInstance(), PermissionManager.FINE_LOCATION)) {
            if (locManager != null) {
                Location location = locManager.getLastKnownLocation(LocationManager.PASSIVE_PROVIDER);
                if (location != null) {
                    callback.onLocationFound(location);
                } else {
                    callback.onError(PocketMap.getInstance().getString(R.string.message_location_not_found));
                }
            }
        } else {
            callback.onPermissionRequired(
                    PermissionManager.FINE_LOCATION,
                    PermissionManager.FIND_MY_LOCATION_REQUEST);
        }
    }

    @Override
    public void subscribeOnLocationChanges(OnLocationSearchResultCallback callback) {
        if (PermissionManager.checkPermission(PocketMap.getInstance(), PermissionManager.FINE_LOCATION)) {
            registerListener(callback);
            if (locManager != null) {
                locManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, REQUEST_MIN_TIME, REQUEST_MIN_DISTANCE, locListener);
            }
        } else
            callback.onPermissionRequired(PermissionManager.FINE_LOCATION, PermissionManager.SUBSCRIBE_ON_LOCATION_UPDATES);
    }

    private void registerListener(final OnLocationSearchResultCallback callback) {
        locListener = new LocationListener() {
            @Override
            public void onLocationChanged(Location location) {
                if (location != null) {
                    callback.onLocationFound(location);
                }
            }

            @Override
            public void onStatusChanged(String s, int i, Bundle bundle) {
                switch (i) {
                    case LocationProvider.AVAILABLE:
                        Log.d("onStatusChanged", "LocationProvider avaliable");
                        break;
                    case LocationProvider.TEMPORARILY_UNAVAILABLE:
                        Log.d("onStatusChanged", "LocationProvider temporarily unavailable");
                        break;
                    case LocationProvider.OUT_OF_SERVICE:
                        callback.onError(PocketMap.getInstance().getString(R.string.message_location_provider_out_of_service));
                        Log.d("onStatusChanged", "LocationProvider out of service");
                        break;
                }
            }

            @Override
            public void onProviderEnabled(String s) {
                Log.d("onProviderEnabled", "LocationProvider enabled");
            }

            @Override

            public void onProviderDisabled(String s) {
                Log.d("onProviderDisabled", "LocationProvider disabled");
            }
        };
    }

    @Override
    public void unsubscribeOfLocationChanges() {
        if (locListener != null) {
            locManager.removeUpdates(locListener);
            locListener = null;
        }
    }
}

File: android2_lesson03/src/kotlin/java/com/example/ama/android2_lesson03/repo/data/location/LocationManagerAndroid.kt
package com.example.ama.android2_lesson03.repo.data.location

import android.content.Context
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.location.LocationProvider
import android.os.Bundle
import android.util.Log
import com.example.ama.android2_lesson03.PocketMap
import com.example.ama.android2_lesson03.R
import com.example.ama.android2_lesson03.repo.SearchQueryManager
import com.example.ama.android2_lesson03.repo.data.base.BaseLocationManager
import com.google.android.gms.maps.model.LatLng

/**
 * Class for locating by android.location
 */
class LocationManagerAndroid : BaseLocationManager() {

    companion object {
        const val MIN_REQUEST_INTERVAL = 2000L
        const val MIN_DISTANCE_REQUEST = 0f
    }

    private val locationManager = PocketMap.instance.getSystemService(Context.LOCATION_SERVICE) as LocationManager
    private var listener: LocationListener? = null

    @SuppressWarnings("MissingPermission")
    override fun findMyLocation(found: (location: Location) -> Unit, notFound: (message: String) -> Unit) {
        val location = locationManager.getLastKnownLocation(LocationManager.PASSIVE_PROVIDER)
        if (location != null) {
            found.invoke(location)
        } else {
            notFound.invoke(PocketMap.instance.getString(R.string.message_location_not_found))
        }
    }

    @SuppressWarnings("MissingPermission")
    override fun subscribeOnLocationUpdates(locationFound: (location: Location) -> Unit, error: (message: String) -> Unit) {
        registerListener(locationFound, error)
        locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, MIN_REQUEST_INTERVAL, MIN_DISTANCE_REQUEST, listener)
    }

    override fun unsubscribeOfLocationUpdates() {
        if (listener != null) {
            locationManager.removeUpdates(listener)
            listener = null
        }
    }

    private fun registerListener(locationFound: (location: Location) -> Unit,
                                 error: (message: String) -> Unit) {
        listener = object : LocationListener {
            override fun onLocationChanged(p0: Location?) {
                if (p0 != null) {
                    locationFound.invoke(p0)
                }
            }

            override fun onStatusChanged(p0: String?, p1: Int, p2: Bundle?) {
                when (p1) {
                    LocationProvider.AVAILABLE -> Log.d("onStatusChanged", "LocationProvider avaliable")
                    LocationProvider.TEMPORARILY_UNAVAILABLE -> Log.d("onStatusChanged", "LocationProvider temporarily unavailable")
                    LocationProvider.OUT_OF_SERVICE -> {
                        error.invoke(PocketMap.instance.getString(R.string.message_location_provider_out_of_service))
                        Log.d("onStatusChanged", "LocationProvider out of service")
                    }
                }
            }

            override fun onProviderEnabled(p0: String?) {
                Log.d("onProviderEnabled", "LocationProvider enabled")
            }

            override fun onProviderDisabled(p0: String?) {
                Log.d("onProviderDisabled", "LocationProvider disabled")
            }
        }
    }
}